### PR TITLE
Move velocity tracker to dart

### PR DESCRIPTION
**Pull request by krisgiesing**
Also adds tests for the least-squares solver and the velocity tracker, and adds a benchmark comparing the native and dart trackers.

The native velocity tracker is left intact for now; it can easily be removed in a separate pass.

### CODE REVIEW

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 1]
```diff
@@ -0,0 +1,190 @@
+import "dart:math" as math;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: Please add a copyright block.

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import "dart:typed_data";
+
+class Vector {
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: Can we use https://www.dartdocs.org/documentation/vector_math/1.4.3/vector_math_64/Vector4-class.html instead?

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+      if (i > 0)
+        result += ", ";
+        result += this[i].toString();
+    }
+    return result;
+  }
+
+  final int _offset;
+  final int _length;
+  final List<double> _elem;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: _elem -> please use complete words in variable names.  _elements?

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+class Matrix {
+  Matrix(int rows, int cols)
+  : _rows = rows,
+    _cols = cols,
+    _elem = new Float64List(rows * cols);
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: Please use complete words in variable names.

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    // Orthonormal basis, column-major ordVectorer.
+    Matrix q = new Matrix(n, m);
+    // Upper triangular matrix, row-major order.
+    Matrix r = new Matrix(n, n);
+    for (int j = 0; j < n; j++) {
+      for (int h = 0; h < m; h++) {
+        q.set(j, h, a.get(j, h));
+      }
+      for (int i = 0; i < j; i++) {
+        double dot = q.getRow(j)*q.getRow(i);
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: Please add a space before and after the `*`

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        }
+      }
+
+      double norm = q.getRow(j).norm();
+      if (norm < 0.000001) {
+        // vectors are linearly dependent or zero so no solution
+        return null;
+      }
+
+      double invNorm = 1.0 / norm;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: invNorm -> inverseNorm ?

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    // SSerr is the residual sum of squares (variance of the error),
+    // and SStot is the total sum of squares (variance of the data) where each
+    // has been weighted.
+    double ymean = 0.0;
+    for (int h = 0; h < m; h++) {
+      ymean += y[h];
+    }
+    ymean /= m;
+
+    double sserr = 0.0;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: sserr -> sumOfSquaredErrors?

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    double sstot = 0.0;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: sstot -> sumOfSquares ?

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+      for (int i = 1; i < n; i++) {
+        term *= x[h];
+        err -= term * out_b[i];
+      }
+      sserr += w[h] * w[h] * err * err;
+      double v = y[h] - ymean;
+      sstot += w[h] * w[h] * v * v;
+    }
+
+    double det = sstot > 0.000001 ? 1.0 - (sserr / sstot) : 1.0;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: det -> determinate?

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 10]
```diff
@@ -0,0 +1,234 @@
+import 'dart:ui' as ui;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: License block please.

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    }
+    return result;
+  }
+
+  final int _offset;
+  final int _length;
+  final List<double> _elem;
+}
+
+class Matrix {
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: Should Matrix and Vector be private to this file?  We probably don't want to cause confusion with our use of Matrix4 elsewhere for transform matricies.

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+  void addMovement(double timeStamp, double x, double y);
+  bool getEstimator(Estimator estimator);
+  void clear();
+}
+
+enum Weighting {
+  WEIGHTING_NONE,
+  WEIGHTING_DELTA,
+  WEIGHTING_CENTRAL,
+  WEIGHTING_RECENT
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: These should be in camelCase

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+class Movement {
+  double event_time = 0.0;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: event_time -> eventTime

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+  ui.Point position = ui.Point.origin;
+}
+
+class LeastSquaresVelocityTrackerStrategy extends VelocityTrackerStrategy {
+  static const int kHistorySize = 20;
+  static const int kHorizonMS = 100;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: Should this be a Duration ?

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+
+  bool getEstimator(Estimator estimator) {
+    // Iterate over movement samples in reverse time order and collect samples.
+    List<double> x = new List<double>();
+    List<double> y = new List<double>();
+    List<double> w = new List<double>();
+    List<double> time = new List<double>();
+    int m = 0;
+    int index = _index;
+    Movement newest_movement = _getMovement(index);
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: newest_movement -> newestMovement

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+    switch (weighting) {
+      case Weighting.WEIGHTING_DELTA:
+        // Weight points based on how much time elapsed between them and the next
+        // point so that points that "cover" a shorter time span are weighed less.
+        //   delta  0ms: 0.5
+        //   delta 10ms: 1.0
+        if (index == _index) {
+          return 1.0;
+        }
+        int next_index = (index + 1) % kHistorySize;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: All these unix_hacker names should be camelCase

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+        result = new Movement();
+        _movements[i] = result;
+      }
+      return result;
+  }
+
+}
+
+class VelocityTracker {
+  static const int kAssumePointerMoveStoppedTimeMs = 40;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: Duration?

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  GestureVelocity getVelocity() {
+    Estimator estimator = new Estimator();
+    if (_strategy.getEstimator(estimator) && estimator.degree >= 1) {
+      // convert from pixels/ms to pixels/s
+      return new GestureVelocity(
+        isValid: true,
+        x: estimator.xcoeff[1]*1000,
+        y: estimator.ycoeff[1]*1000
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: Spaces around `*` pls

File: sky/packages/sky/lib/src/gestures/velocity_tracker.dart [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+      if (result == null) {
+        result = new Movement();
+        _movements[i] = result;
+      }
+      return result;
+  }
+
+}
+
+class VelocityTracker {
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/velocity_tracker.dart]: I wonder how many of these classes should be private.  I'd be inclined to make them all private except this one.

File: sky/packages/sky/lib/src/widgets/event_recorder.dart [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+import 'basic.dart';
+import 'framework.dart';
+
+enum EventRecorderMode {
+  stop,
+  record
+}
+
+typedef void EventsReady(List<PointerInputEvent> events);
```
Comment (abarth) [inline sky/packages/sky/lib/src/widgets/event_recorder.dart]: We usually use the suffix "Callback" for these types.  Not sure if having that suffix is all that valuable, but there's some value in being consistent.

File: sky/packages/sky/lib/src/widgets/event_recorder.dart [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+    _mode = config.mode;
+  }
+
+  void didUpdateConfig(EventRecorder oldConfig) {
+    if (_mode == EventRecorderMode.record &&
+        config.mode == EventRecorderMode.stop) {
+      config.onEventsReady(_events);
+      _events.clear();
+    }
+    _mode = config.mode;
```
Comment (abarth) [inline sky/packages/sky/lib/src/widgets/event_recorder.dart]: There's no need to store the _mode on `this`.  You can check the old value of mode from `oldConfig.mode`.

**Comment by abarth** [comment]
LGTM

The above comments are all just style nits.

**Comment by j9brown** [comment]
BTW, I wrote some (not great) tests for the velocity tracker in CTS,
including a few that I had to disable because I couldn't come up with an
algorithm to pass them...  Stretch goal?  :)

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart (hunk 2 above)
Comment (krisgiesing) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: Yup, that's the issue.

File: sky/packages/sky/lib/src/gestures/lsq_solver.dart (hunk 11 above)
Comment (krisgiesing) [inline sky/packages/sky/lib/src/gestures/lsq_solver.dart]: Probably so.

**Comment by krisgiesing** [comment]
I should have mentioned previously, the variable names and some style idioms were preserved exactly as they were in the C++ implementation so I could more easily do visual comparisons. Now that the Dart implementation appears to conform, I can go back and do a coding style pass.

File: sky/packages/sky/lib/src/widgets/event_recorder.dart (hunk 21 above)
Comment (krisgiesing) [inline sky/packages/sky/lib/src/widgets/event_recorder.dart]: I wasn't sure whether config.mode would always be available for checking during the event processing.

File: sky/packages/sky/lib/src/widgets/event_recorder.dart (hunk 20 above)
Comment (krisgiesing) [inline sky/packages/sky/lib/src/widgets/event_recorder.dart]: Will change, thanks

File: sky/packages/sky/lib/src/widgets/event_recorder.dart (hunk 21 above)
Comment (abarth) [inline sky/packages/sky/lib/src/widgets/event_recorder.dart]: Yep, it will.

**Comment by krisgiesing** [comment]
Ugh, how did I end up committing all this other stuff?

**Comment by krisgiesing** [comment]
I authored these commits.

**Pull request merged**